client/main/options.js

```diff
@@ -2,166 +2,178 @@
+    });
+
+    this._override('getUsedColumns', (baseFunction) => {
+        let r = [];
+        if (this.children) {
+            for (let i = 0; i < this.children.length; i++) {
+                let child = this.children[i];
+                r = r.concat(child.getUsedColumns());
             }
         }
-        else {
-            value = new Array(options.options.length);
-            for (let j = 0; j < options.options.length; j++) {
-                let option = options.options[j];
-                value[j] = _fromPB(option, Messages);
+        return r;
+    });
+};
+SuperClass.create(OptionType.ArrayType);
+
+OptionType.GroupType = function(template, value) {
+    OptionType.Option.extendTo(this, template, value);
+
+    this._override('update', (baseFunction, value) => {
+        baseFunction.call(this, value);
+        this.children = [];
+        if (value === null)
+            return;
+        for (let i = 0; i < template.elements.length; i++) {
+            let element = template.elements[i];
+            this.children.push(OptionType.create(element, value[element.name]));
+        }
+    });
+
+    this._override('getUsedColumns', (baseFunction) => {
+        let r = [];
+        if (this.children) {
+            for (let i = 0; i < this.children.length; i++) {
+                let child = this.children[i];
+                r = r.concat(child.getUsedColumns());
             }
         }
-    }
-    else if (option.type === 'o') {
-        if (option.o === Messages.AnalysisOption.Other.TRUE)
-            value = true;
-        else if (option.o === Messages.AnalysisOption.Other.FALSE)
-            value = false;
-        else
-            value = null;
-    }
-    else {
-        value = option[option.type];
-    }
-
-    return value;
+        return r;
+    });
 };
+SuperClass.create(OptionType.GroupType);
 
-Options.fromPB = function(optionsPB, Messages) {
 
-    let value = { };
 
-    for (let j = 0; j < optionsPB.names.length; j++) {
-        let name = optionsPB.names[j];
-        let option = optionsPB.options[j];
-        value[name] = _fromPB(option, Messages);
+const Options = function(def) {
+
+    this._options = {};
+
+    for (var i = 0; i < def.length; i++) {
+        var template = def[i];
+        var option = OptionType.create(template, template.default);
+        this._options[template.name] = option;
     }
 
-    return value;
+    this.getUsedColumns = function() {
+        let r = [];
+        for (let name in this._options) {
+            let option = this._options[name];
+            r = r.concat(option.getUsedColumns());
+        }
+        r = _.uniq(r);
+        return r;
+    };
+
+    this.getOption = function(name) {
+        return this._options[name];
+    };
+
+    this.setValues = function(values) {
+        for (let name in values) {
+            if (name in this._options) {
+                if (values[name] !== undefined)
```

Comment:
is this if statement necessary?
will the for loop `for (let name in values)` ever produce a `values[name]` value which is undefined?